File: base/src/main/groovy/com/knight/plugin/doublecheck/transform/KnightTransform.java
package com.knight.plugin.doublecheck.transform;

import com.android.build.api.transform.Context;
import com.android.build.api.transform.DirectoryInput;
import com.android.build.api.transform.Format;
import com.android.build.api.transform.JarInput;
import com.android.build.api.transform.QualifiedContent;
import com.android.build.api.transform.Status;
import com.android.build.api.transform.Transform;
import com.android.build.api.transform.TransformException;
import com.android.build.api.transform.TransformInput;
import com.android.build.api.transform.TransformInvocation;
import com.android.build.api.transform.TransformOutputProvider;
import com.android.build.gradle.internal.LoggerWrapper;
import com.android.build.gradle.internal.pipeline.TransformManager;
import com.android.ide.common.internal.WaitableExecutor;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Created by Quinn on 26/02/2017.
 * Transform to modify bytecode
 */
public class KnightTransform extends Transform {

    private static final LoggerWrapper logger = LoggerWrapper.getLogger(KnightTransform.class);


    protected BaseWeaver bytecodeWeaver;
    private WaitableExecutor waitableExecutor;

    public KnightTransform() {
        this.waitableExecutor = WaitableExecutor.useGlobalSharedThreadPool();
    }

    @Override
    public String getName() {
        return "DoubleCheckTransform";
    }

    @Override
    public Set<QualifiedContent.ContentType> getInputTypes() {
        return TransformManager.CONTENT_CLASS;
    }

    @Override
    public Set<QualifiedContent.Scope> getScopes() {
        return TransformManager.SCOPE_FULL_PROJECT;
    }

    @Override
    public boolean isIncremental() {
        return true;
    }


    @Override
    public void transform(TransformInvocation transformInvocation) throws TransformException, InterruptedException, IOException {
        long startTime = System.currentTimeMillis();
        boolean isIncremental = transformInvocation.isIncremental();
        TransformOutputProvider outputProvider = transformInvocation.getOutputProvider();
        Collection<TransformInput> inputs = transformInvocation.getInputs();

        if (!isIncremental) {
            outputProvider.deleteAll();
        }
        for (TransformInput input : inputs) {
            for (JarInput jarInput : input.getJarInputs()) {
                Status status = jarInput.getStatus();
                File dest = outputProvider.getContentLocation(
                        jarInput.getFile().getAbsolutePath(),
                        jarInput.getContentTypes(),
                        jarInput.getScopes(),
                        Format.JAR);
                if (isIncremental) {
                    switch (status) {
                        case NOTCHANGED:
                            break;
                        case ADDED:
                        case CHANGED:
                            transformJar(jarInput.getFile(), dest, status);
                            break;
                        case REMOVED:
                            if (dest.exists()) {
                                FileUtils.forceDelete(dest);
                            }
                            break;
                    }
                } else {
                    transformJar(jarInput.getFile(), dest, status);
                }
            }

            for (DirectoryInput directoryInput : input.getDirectoryInputs()) {
                File dest = outputProvider.getContentLocation(directoryInput.getName(),
                        directoryInput.getContentTypes(), directoryInput.getScopes(),
                        Format.DIRECTORY);
                FileUtils.forceMkdir(dest);
                if (isIncremental) {
                    String srcDirPath = directoryInput.getFile().getAbsolutePath();
                    String destDirPath = dest.getAbsolutePath();
                    Map<File, Status> fileStatusMap = directoryInput.getChangedFiles();
                    for (Map.Entry<File, Status> changedFile : fileStatusMap.entrySet()) {
                        Status status = changedFile.getValue();
                        File inputFile = changedFile.getKey();
                        String destFilePath = inputFile.getAbsolutePath().replace(srcDirPath, destDirPath);
                        File destFile = new File(destFilePath);
                        switch (status) {
                            case NOTCHANGED:
                                break;
                            case REMOVED:
                                if (destFile.exists()) {
                                    //noinspection ResultOfMethodCallIgnored
                                    destFile.delete();
                                }
                                break;
                            case ADDED:
                            case CHANGED:
                                FileUtils.touch(destFile);
                                transformSingleFile(inputFile, destFile, srcDirPath);
                                break;
                        }
                    }
                } else {
                    transformDir(directoryInput.getFile(), dest);
                }

            }

        }

        waitableExecutor.waitForTasksWithQuickFail(true);
        long costTime = System.currentTimeMillis() - startTime;
        logger.info((getName() + " costed " + costTime + "ms"));
    }

    private void transformSingleFile(final File inputFile, final File outputFile, final String srcBaseDir) {
        waitableExecutor.execute(() -> {
            bytecodeWeaver.weaveSingleClassToFile(inputFile, outputFile, srcBaseDir);
            return null;
        });
    }

    private void transformDir(final File inputDir, final File outputDir) throws IOException {
        final String inputDirPath = inputDir.getAbsolutePath();
        final String outputDirPath = outputDir.getAbsolutePath();
        if (inputDir.isDirectory()) {
            for (final File file : com.android.utils.FileUtils.getAllFiles(inputDir)) {
                waitableExecutor.execute(() -> {
                    String filePath = file.getAbsolutePath();
                    File outputFile = new File(filePath.replace(inputDirPath, outputDirPath));
                    bytecodeWeaver.weaveSingleClassToFile(file, outputFile, inputDirPath);
                    return null;
                });
            }
        }
    }

    private void transformJar(final File srcJar, final File destJar, Status status) {
        waitableExecutor.execute(() -> {
            bytecodeWeaver.weaveJar(srcJar, destJar);
            return null;
        });
    }
}

File: knight-transform/src/main/kotlin/com/knight/transform/KnightTransform.kt
package transform

import com.android.build.api.transform.*
import com.android.build.gradle.AppPlugin
import com.android.build.gradle.LibraryPlugin
import com.android.build.gradle.internal.LoggerWrapper
import com.android.build.gradle.internal.pipeline.TransformManager
import com.android.ide.common.internal.WaitableExecutor
import transform.asm.BaseWeaver
import org.apache.commons.io.FileUtils
import org.gradle.api.Project
import java.io.File

open abstract class KnightTransform(project: Project) : Transform() {
    private val logger = LoggerWrapper.getLogger(javaClass::class.java)

    private val waitableExecutor = WaitableExecutor.useGlobalSharedThreadPool()

    protected lateinit var baseWeaver: BaseWeaver

    protected val isLibrary = project.plugins.hasPlugin(LibraryPlugin::class.java)

    override fun getName(): String {
        return this.javaClass.simpleName
    }

    override fun getInputTypes(): MutableSet<QualifiedContent.ContentType> {
        return TransformManager.CONTENT_CLASS
    }

    override fun isIncremental(): Boolean {
        return true
    }

    override fun getScopes(): MutableSet<in QualifiedContent.Scope> {
        return if (isLibrary) TransformManager.PROJECT_ONLY else TransformManager.SCOPE_FULL_PROJECT
    }

    open fun isNeedScanJar(): Boolean {
        return true
    }


    override fun transform(transform: TransformInvocation) {
        val startTime = System.currentTimeMillis()
        transform.apply {
            if (!isIncremental) outputProvider.deleteAll()
            inputs.forEach {
                it.jarInputs.forEach { jarInput ->
                    val status = jarInput.status
                    val dest = outputProvider.getContentLocation(
                            jarInput.file.absolutePath,
                            jarInput.contentTypes,
                            jarInput.scopes,
                            Format.JAR)
                    if (isIncremental) {
                        when (status) {
                            Status.ADDED, Status.CHANGED -> {
                                transformJar(jarInput.file, dest, isNeedScanJar())
                            }
                            Status.REMOVED -> {
                                if (dest.exists()) {
                                    FileUtils.forceDelete(dest)
                                }
                            }
                            else -> {
                            }
                        }
                    } else {
                        transformJar(jarInput.file, dest, isNeedScanJar())
                    }
                }

                it.directoryInputs.forEach { directoryInput ->
                    val dest = outputProvider.getContentLocation(
                            directoryInput.name,
                            directoryInput.contentTypes,
                            directoryInput.scopes,
                            Format.DIRECTORY)
                    FileUtils.forceMkdir(dest)
                    if (isIncremental) {
                        val srcDirPath = directoryInput.file.absolutePath
                        val destDirPath = dest.absolutePath
                        directoryInput.changedFiles.forEach { inputFile, status ->
                            val destFilePath = inputFile.absolutePath.replace(srcDirPath, destDirPath)
                            val destFile = File(destFilePath)
                            when (status) {
                                Status.ADDED, Status.CHANGED -> {
                                    FileUtils.touch(destFile)
                                    transformSingleFile(inputFile, destFile, srcDirPath)
                                }
                                Status.REMOVED -> {
                                    if (destFile.exists()) {
                                        destFile.delete()
                                    }
                                }
                                else -> {
                                }
                            }
                        }
                    } else {
                        transformDir(directoryInput.file, dest)
                    }

                }
            }
        }
        waitableExecutor.waitForTasksWithQuickFail<Any>(true)
        val costTime = System.currentTimeMillis() - startTime
        logger.info(name + " costed " + costTime + "ms")
        println(name + " costed " + costTime + "ms")
    }

    private fun transformJar(srcJar: File, destJar: File, isNeedScan: Boolean) {
        waitableExecutor.execute {
            if (isNeedScan) {
                baseWeaver.weavJar(srcJar, destJar)
            } else {
                FileUtils.copyFile(srcJar, destJar)
            }
        }
    }

    private fun transformSingleFile(inputFile: File, outputFile: File, srcDir: String) {
        waitableExecutor.execute {
            baseWeaver.weaveFile(inputFile, outputFile, srcDir)
        }
    }

    private fun transformDir(inputFile: File, outputFile: File) {
        val inputDirPath = inputFile.getAbsolutePath()
        val outputDirPath = outputFile.getAbsolutePath()
        if (inputFile.isDirectory) {
            com.android.utils.FileUtils.getAllFiles(inputFile).forEach {
                waitableExecutor.execute {
                    val filePath = it.absolutePath
                    val outputFile = File(filePath.replace(inputDirPath, outputDirPath))
                    baseWeaver.weaveFile(it, outputFile, inputDirPath)
                }
            }
        }
    }

}